### PR TITLE
Make spark_udf use NFS to broadcast model to spark executor on databricks runtime and spark connect mode

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/10463?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10463/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10463
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Make spark_udf use NFS to broadcast model to spark executor on databricks runtime and spark connect mode.

If the client code is running in databricks runtime (e.g. running in a databricks notebook), then we can use NFS directory (e.g. dbutils.entry_point.getReplNFSTempDir()) to broadcast local model files to executor side, note databricks shared cluster also supporting this, the NFS directory is mounted to the same path in safe spark UDF container. 

Testing this on databricks shared cluster:
https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/1841156755475244/command/1841156755475277

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by WeichenXu123** [comment]
@mlflow-automation autoformat

File: mlflow/pyfunc/__init__.py [hunk 1]
```diff
@@ -1516,6 +1529,8 @@ def udf(
                     model_uri=local_model_path_on_executor, capture_output=True
                 )
             else:
+                if is_spark_connect:
+                    assert should_spark_connect_use_nfs
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1533]: can we remove this?

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -1239,10 +1239,7 @@ def spark_udf(
     from mlflow.utils._spark_utils import _SparkDirectoryDistributor
 
     is_spark_connect = _is_spark_connect()
-    if is_spark_connect and env_manager in (_EnvManager.VIRTUALENV, _EnvManager.CONDA):
-        raise MlflowException.invalid_parameter_value(
-            f"Environment manager {env_manager!r} is not supported in Spark connect mode.",
-        )
+
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1242]:
```suggestion
```

File: mlflow/pyfunc/__init__.py [hunk 3]
```diff
@@ -1260,6 +1257,22 @@ def spark_udf(
...
+    # then in executor side we can load model from the NFS temp directory.
+    should_spark_connect_use_nfs = is_in_databricks_runtime() and should_use_nfs
+
+    if (
+        is_spark_connect
+        and env_manager in (_EnvManager.VIRTUALENV, _EnvManager.CONDA)
+        and not should_spark_connect_use_nfs
+    ):
+        raise MlflowException.invalid_parameter_value(
+            f"In non-databricks runtime or NFS is not available, "
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1272]: what is non-databricks runtime?

File: mlflow/pyfunc/__init__.py [hunk 4]
```diff
@@ -1581,16 +1596,19 @@ def batch_predict_fn(pdf, params=None):
...
-                try:
-                    loaded_model = mlflow.pyfunc.load_model(model_path)
-                except Exception:
-                    os.makedirs(model_path, exist_ok=True)
-                    loaded_model = mlflow.pyfunc.load_model(model_uri, dst_path=model_path)
+                if should_spark_connect_use_nfs:
+                    loaded_model = mlflow.pyfunc.load_model(local_model_path)
+                else:
+                    model_path = os.path.join(
+                        tempfile.gettempdir(),
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1603]: I think we should use `create_tmp_dir` in case the model is large.

File: mlflow/pyfunc/__init__.py (hunk 3 above)
Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:1272]: e.g., user runs spark connect client code on their own laptop to connect to databricks shared cluster.

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:1272]: is the error message good ? then I will update test to match new error message

File: mlflow/pyfunc/__init__.py [hunk 5]
```diff
@@ -1581,16 +1594,19 @@ def batch_predict_fn(pdf, params=None):
 
         elif env_manager == _EnvManager.LOCAL:
             if is_spark_connect:
```
Comment (serena-ruan) [inline mlflow/pyfunc/__init__.py:1596]: If we update the condition to be `is_spark_connect and not should_spark_connect_use_nfs` then we don't need below if else statement. But I'm fine keeping it here as well :)

File: mlflow/pyfunc/__init__.py [hunk 6]
```diff
@@ -1260,6 +1257,22 @@ def spark_udf(
...
+    should_spark_connect_use_nfs = is_in_databricks_runtime() and should_use_nfs
+
+    if (
+        is_spark_connect
+        and env_manager in (_EnvManager.VIRTUALENV, _EnvManager.CONDA)
+        and not should_spark_connect_use_nfs
+    ):
+        raise MlflowException.invalid_parameter_value(
+            f"In non-databricks runtime or NFS is not available, "
+            f"environment manager {env_manager!r} is not supported in Spark connect mode.",
```
Comment (serena-ruan) [inline mlflow/pyfunc/__init__.py:1273]: /ai Could you improve the error message to be more readable?

Comment (serena-ruan) [reply mlflow/pyfunc/__init__.py:1273]: Maybe "Environment manager {env_manager!r} is not supported in Spark connect mode when in non-databricks runtime or NFS not available" ?

File: mlflow/pyfunc/__init__.py [hunk 7]
```diff
@@ -1260,6 +1257,22 @@ def spark_udf(
...
+    should_spark_connect_use_nfs = is_in_databricks_runtime() and should_use_nfs
+
+    if (
+        is_spark_connect
+        and env_manager in (_EnvManager.VIRTUALENV, _EnvManager.CONDA)
+        and not should_spark_connect_use_nfs
+    ):
+        raise MlflowException.invalid_parameter_value(
+            f"Environment manager {env_manager!r} is not supported in Spark connect mode when "
+            "in non-databricks runtime or NFS not available",
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1273]:
```suggestion
            "in non-databricks environment or NFS not available",
```

Comment (harupy) [inline mlflow/pyfunc/__init__.py:1273]: /ai fix grammatical errors

File: mlflow/pyfunc/__init__.py [hunk 8]
```diff
@@ -1260,6 +1257,22 @@ def spark_udf(
         is_spark_in_local_mode or should_use_nfs or is_spark_connect
     )
 
+    # For spark connect mode,
+    # If client code is executed in databricks runtime and NFS is available,
+    # In driver side, we save model to NFS temp directory,
+    # then in executor side we can load model from the NFS temp directory.
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1263]:
```suggestion
    # we save model to NFS temp directory in the driver
    # and load the model in the executor.
```

not sure if this makes sense.

**Review by harupy — APPROVED** [review]

File: mlflow/pyfunc/__init__.py [hunk 9]
```diff
@@ -1580,9 +1593,9 @@ def batch_predict_fn(pdf, params=None):
                 return client.invoke(pdf).get_predictions()
 
         elif env_manager == _EnvManager.LOCAL:
-            if is_spark_connect:
+            if is_spark_connect and not should_spark_connect_use_nfs:
                 model_path = os.path.join(
-                    tempfile.gettempdir(),
+                    create_tmp_dir(),
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1598]: Does this create a unique directory in each invocation?

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py]: Good catch. For non-databricks environment, it returns different path each invocation.
I move this path generation to driver side.

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py]: emm, moving to driver side might have other issue, driver side might has different OS with executor side, then legal tmp path in driver side might not be available in executor side. e.g. macOS returns temp path like `'/var/folders/z7/47xgtbq53kg8hs1zn7qk8kw40000gp/T/tmptdo5n4l5'`

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py]: I will revert it to `tempfile.gettempdir()` for now.

**Review by serena-ruan — APPROVED** [review]
LGTM!

File: mlflow/pyfunc/__init__.py (hunk 9 above)
Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py]: This branch code will be removed once `spark_session.add_artifacts` API is ready in databricks shared cluster. See https://databricks.atlassian.net/browse/ML-36703

**Pull request merged**